src/main/groovy/com/lesfurets/jenkins/unit/declarative/StageDeclaration.groovy

```diff
@@ -41,24 +63,42 @@ class StageDeclaration extends GenericPipelineDeclaration {
     }
 
     def execute(Object delegate) {
+        Map envValuestoRestore = [:]
+
         String name = this.name
         this.options.each {
-            executeOn(delegate, it)
+            executeWith(delegate, it)
         }
         if(parallel) {
             parallel.execute(delegate)
         }
 
-        if(delegate.binding.variables.currentBuild.result == "FAILURE"){
+        if (delegate.currentBuild.result == "FAILURE") {
             executeWith(delegate, { echo "Stage \"$name\" skipped due to earlier failure(s)" })
             return
         }
 
         if (!when || when.execute(delegate)) {
-            super.execute(delegate)
+            // Set environment for stage
```

Comment:
I would prefer not to copy-paste same code in different classes inheritance here is a good practice

Reply:
The stage should clear the environment/params after finishing so the next stage could inherit pipeline vars. But i guess we could implement the cleanup at GenericPipeline level

Reply:
I created a public static method to init and reset the environment props. This way we can share the logic, but choose when to call the parts.

We want to init the env in a stage, but only reset it after the stage is complete. Super.execute would have immediately cleared the stage custom env overrides.